Add a built-in debug overlay that displays current performance

Calinou opened the pull request:
This can be used to diagnose performance issues, both for developers and players (since this is also available in exported projects). Compared to the editor's **Monitors** tab, this has a few advantages:

- Performance information is overlaid on the running game, so you can still view it while in fullscreen on a single-monitor setup.
- The editor doesn't need to be running in the background, so it works in exported projects where performance is the most indicative of what players will experience.

Press <kbd>F3</kbd> to cycle between 3 modes: hidden, compact (FPS/frametime display only), detailed. This key can be changed in the project settings' input map by modifying `ui_cycle_debug_overlay_display_mode` (remember to enable **Show Built-in Actions**). The debug overlay can also be programmatically shown by changing `SceneTree.debug_overlay_mode` from a script.

The debug overlay's scale automatically adapts to the window size if using the `canvas_items` or `viewport` stretch mode. The default scale is controlled using the `gui/theme/default_theme_scale` project setting.

This PR is a C++ port of the [godot-debug-menu](https://github.com/godot-extended-libraries/godot-debug-menu) add-on I wrote with a few tweaks. Happy holidays :slightly_smiling_face:

- This closes https://github.com/godotengine/godot-proposals/issues/63.

## Preview

### Forward+

*Example integration in the game editor (notice how the viewport sizes match):*

![Forward+](https://github.com/user-attachments/assets/436e00eb-1070-436f-8e0f-80a3b18aa68d)

### Mobile

![Mobile](https://github.com/user-attachments/assets/8c9e310b-053b-4119-9795-b0fd0c01f1e4)

### Compatibility

*Featuring a shader compilation stutter, so you can see what it looks like on the graph.*

![Compatibility](https://github.com/user-attachments/assets/1e70f4f8-ddb1-4cb9-a128-ce813d4b8938)

## TODO

- [x] Rename to something else than DebugMenu, e.g. PerformanceOverlay. This "menu" doesn't provide any options you can interact with and it's not really related to *debugging* per se, so a different name would be more suited.
  - Renamed to DebugOverlay.
- [ ] Add a project setting to set the debug overlay mode to use on startup.
- [ ] Add a project setting to change the location the debug overlay appears in? This could be needed for some games where it'd overlap with HUD elements.
  - This could be an enum with 8 positions (top-left, top center, top-right, and so on).
- [ ] Update the graphics settings description whenever any of the graphics settings is changed in RenderingServer. This can likely be done by emitting a signal in RenderingServer and connecting to it in DebugOverlay.
- [ ] Make the graphics settings description more compact somehow? While I've made the C++ port more compact than the add-on by reducing line spacing, it still takes up quite a bit of screen real estate right now.
  - One way to make it more compact is to write abbreviations next to each other, similar to what DuckStation and PCSX2 do with their debug overlays:

![image](https://github.com/user-attachments/assets/b5c8c7db-bd74-4aa0-ba83-87734ef761ab)

- [ ] Prevent DebugOverlay nodes from being selectable in the game editor (as they can get in the way). @YeldhamDev Is there some kind of node metadata available for this?

![Debug overlay in game editor](https://github.com/user-attachments/assets/beefc5b7-c851-4f61-9518-71c441584ec0)

- [ ] Figure out the occasional frametime spikes that occur with the detailed view. This is most likely due to graph drawing as commenting it out fixes the issue. Perhaps switching to Image-based drawing like the editor (visual) profiler might work better.
  - This is not a recent engine regression nor an issue specific to the C++ port as the [add-on](https://github.com/godot-extended-libraries/godot-debug-menu) has the same issue, even on 4.2.stable.

Calinou committed:
Add a built-in debug overlay that displays current performance

This can be used to diagnose performance issues, both for developers
and players (since this is also available in exported projects).

Press F3 to cycle between 3 modes: hidden, compact (FPS/frametime display only),
detailed.

drcucaracha commented:
Hi, could you add something like a label to know how much video memory and RAM is consumed by the application/game?

tetrapod00 commented:
I'm not sure it makes sense to color-code the entire bottom section of text based on the currently used rendering method. That same range of color hues (blue, green, reddish) is already used to indicate performance elsewhere in the UI. I think colorcoding just the renderer name or the line it is on might make more sense.

RedMser commented:
While this feature will be largely useful to many people, I don't really agree with the way it is implemented here. It feels very inflexible and too tightly coupled to SceneTree.

Wouldn't it be enough to expose the `DebugMenu` class (or however it'll be named) as a node you can add to your game?
This way you have full control over when it is visible, you can export properties on it which can be modified (e.g. styling, turning display mode into multiple booleans, etc.) and you could even insert it into multiple different viewports at once for split-screen or multi-window games (currently just the main window can have the overlay visible it seems).
For simple cases, the user may just add it as an Autoload so it hooks to the main window like in the current implementation.

This means the SceneTree API, the input map changes and the ProjectSettings changes could all be reverted. The class is exposed to ClassDB. Some logic that currently assumes that the node never moves would also need updating (e.g. detecting reparenting for `viewport_node size_changed` signal).

And in turn it would then require the user to add the node to an appropriate location by hand, and to use scripting to show/hide the overlay instead of a magic built-in action.

Calinou commented:
> Wouldn't it be enough to expose the DebugMenu class (or however it'll be named) as a node you can add to your game?

I've thought about it, but then it's no longer automatically available in every project out of the box. This is an issue when testing MRPs or working on prototypes, as having to configure it manually in every of these projects would get tedious quickly.

There might be an argument for splitscreen but from what I've seen, all these debug overlays are common across all viewports in other engines/games. The information they report doesn't change from viewport to viewport after all, with the exception of resolution (but in splitscreen games, it's often the same across all viewports anyway - just a divisor of the fullscreen size).

Note that the node is dynamically added by SceneTree, so there is no performance or memory cost when the debug menu isn't visible (it's freed once you hide it).

> I'm not sure it makes sense to color-code the entire bottom section of text based on the currently used rendering method. That same range of color hues (blue, green at least, arguably the vaguely red color for Mobile) is already used to indicate performance elsewhere in the UI. I think colorcoding just the renderer name or the line it is on might make more sense.

That can be done, it'll just need splitting the rendering method line to its own node.

> Hi, could you add something like a label to know how much video memory and RAM is consumed by the application/game?

This needs https://github.com/godotengine/godot-proposals/issues/4871 to be implemented first, along with another API to query the amount of VRAM used by the game.

Regarding RAM utilization, Godot has some debug counters available for this, but in my experience, it doesn't accurately reflect the RAM used by the whole Godot process (I'm not sure why). I would need to make sure what it's reporting is 100% accurate before we expose this information in the debug menu, as displaying misleading information is worse than not displaying any information.

arkology commented:
The only thing I know for sure - this one will be great addition for Game Editor/Workspace. Other questions (should it be separate node accessable by users (including in exported release builds), which data it should provide and so on) are still opened.

YeldhamDev commented:
> Prevent DebugMenu nodes from being selectable in the game editor (as they can get in the way). @YeldhamDev Is there some kind of node metadata available for this?

Not currently, no. If such system would be implemented, it would also need to prevent it appearing in the remote tree as well.

timoschwarzer commented:
It would be nice if this would be somehow expandable by user code.
 (`DebugMenuPlugin`?). I'm thinking of stuff like network latency for other games or general debug information that is specific to the game (scene management, 3rd party integrations like FMOD etc.pp.).
In split screen games (i.e. subviewports) the time it took to render that specific viewport would be interesting.

ckaiser commented:
Tested on Windows, works as intended. I've been using your extension on most of my projects so having this built in would be great! 🥳 

Something I noticed but that is purely personal preference is that I like the old colors a bit better, it looks more contrasty and the green seems to over-highlight text that really isn't that important and doesn't change much.

![extension_v_inbuilt](https://github.com/user-attachments/assets/955f5c39-0fd6-4d27-b7d7-370e68e12263)

Alexofp commented:
![pic](https://github.com/user-attachments/assets/8a21043d-9f6a-49ef-8c2f-70ee368e68b4)

Great plugin, would be incredibly useful to have in the engine, I would love if there was a way to show how many drawcalls, gpu memory and triangles are being pushed to the renderer. Very important info when you're trying to optimize your game
I modified the gdscript plugin to show that info and it's been perfect for me. Would be nice to have these options (maybe optionally) somewhere here if it's gonna be added..

Mickeon reviewed (commented):
Docs are ffffffffine

Calinou commented:
> I would love if there was a way to show how many drawcalls, gpu memory and triangles are being pushed to the renderer. Very important info when you're trying to optimize your game

This information is only reported for 3D currently (it doesn't take anything 2D into account), so it would be irrelevant for 2D projects. If it was to be added, it'd have to be a project setting you manually enable, or we could hide it if the reported draw call/primitive count is 0.

Calinou commented on scene/main/scene_tree.cpp:
```diff
@@ -1679,6 +1680,37 @@ bool SceneTree::is_multiplayer_poll_enabled() const {
 	return multiplayer_poll;
 }
 
+void SceneTree::set_debug_menu_display_mode(DebugMenuDisplayMode p_mode) {
```
It's allowed in release builds on purpose (see OP for rationale) :slightly_smiling_face:

fire commented:
The Github action errors are a bit confusing.

```
In file included from ./scene/3d/node_3d.h:35,
                 from ./scene/3d/camera_3d.h:34,
                 from scene/gui/debug_menu.cpp:37:
Error: ./scene/resources/3d/world_3d.h:87:9: error: 'PhysicsDirectSpaceState3D' does not name a type
   87 |         PhysicsDirectSpaceState3D *get_direct_space_state();
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~
scene/gui/debug_menu.cpp: In member function 'void DebugMenu::_update_settings_label()':
Error: scene/gui/debug_menu.cpp:402:29: error: 'const class Window' has no member named 'get_camera_3d'; did you mean 'get_camera_2d'?
  402 |                 if (window->get_camera_3d()) {
      |                             ^~~~~~~~~~~~~
      |                             get_camera_2d
Error: scene/gui/debug_menu.cpp:445:70: error: 'const class Window' has no member named 'get_camera_3d'; did you mean 'get_camera_2d'?
  445 |                         const Ref<Environment> environment = window->get_camera_3d()->get_world_3d()->get_environment();
      |                                                                      ^~~~~~~~~~~~~
      |                                                                      get_camera_2d
scons: *** [scene/gui/debug_menu.linuxbsd.template_release.x86_64.o] Error 1
scons: building terminated because of errors.
```

Calinou commented:
@fire The build error occurs in a minimal template with all modules disabled, because I forgot a `#ifndef 2D_DISABLED` check. I went and added it.

KoBeWi commented:
The text might need bigger outline or some background. It's now always readable.
![image](https://github.com/user-attachments/assets/87e593e7-c65b-4376-96f8-348cf07a2f37)

Calinou commented:
>  Also the game freezes noticeably when enabling the overlay.

This is likely because getting the driver version is slow on Linux. [It's cached for subsequent calls](https://github.com/godotengine/godot/blob/master/platform/linuxbsd/os_linuxbsd.cpp#L309-L312) but the first call can easily take 500+ ms. I remember seeing an issue about this, but I can't find it right now.

KoBeWi commented:
> This is likely because getting the driver version is slow on Linux

I'm on Windows though.
I tested in dev build, but that's still very long freeze.

Calinou commented:
Rebased and tested again, it works as expected. I renamed the class to DebugOverlay as per https://github.com/godotengine/godot/pull/100829#issuecomment-2599043237.

AThousandShips commented on doc/classes/SceneTree.xml:
```diff
@@ -373,5 +376,17 @@
+			Built-in debug overlay is hidden.
+		</constant>
+		<constant name="DEBUG_OVERLAY_DISPLAY_MODE_COMPACT" value="1" enum="DebugOverlayDisplayMode">
+			Built-in debug overlay is visible in compact mode (only FPS, frametime, V-Sync mode and framerate cap displayed).
```
```suggestion
			Built-in debug overlay is visible in compact mode (only FPS, frametime, V-Sync mode, and framerate cap displayed).
```

AThousandShips commented on scene/gui/debug_overlay.cpp:
```diff
@@ -0,0 +1,846 @@
+	fps_graph_panel->set_mouse_filter(Control::MOUSE_FILTER_IGNORE);
+	fps_graph_panel->set_v_size_flags(Control::SIZE_SHRINK_BEGIN);
+	fps_graph_panel->set_custom_minimum_size(Vector2(GRAPH_SIZE_X, GRAPH_SIZE_Y) * default_base_scale);
+	fps_graph_panel->add_theme_style_override("panel", graph_background);
```
```suggestion
	fps_graph_panel->add_theme_style_override(SceneStringName(panel), graph_background);
```
For all

AThousandShips commented on scene/gui/debug_overlay.cpp:
```diff
@@ -0,0 +1,846 @@
+	fps_graph_panel->set_v_size_flags(Control::SIZE_SHRINK_BEGIN);
+	fps_graph_panel->set_custom_minimum_size(Vector2(GRAPH_SIZE_X, GRAPH_SIZE_Y) * default_base_scale);
+	fps_graph_panel->add_theme_style_override("panel", graph_background);
+	fps_graph_panel->connect("draw", callable_mp(this, &DebugOverlay::_graph_draw).bind(GRAPH_TYPE_FPS));
```
```suggestion
	fps_graph_panel->connect(SceneStringName(draw), callable_mp(this, &DebugOverlay::_graph_draw).bind(GRAPH_TYPE_FPS));
```
For all

AThousandShips commented on scene/gui/debug_overlay.h:
```diff
@@ -0,0 +1,166 @@
+#include "scene/gui/label.h"
+#include "scene/gui/margin_container.h"
+#include "scene/gui/panel_container.h"
+#include "scene/gui/rich_text_label.h"
```
Many of these can be forward declared

AThousandShips commented on scene/gui/debug_overlay.h:
```diff
@@ -0,0 +1,166 @@
+	PackedFloat32Array frame_history_cpu;
+	PackedFloat32Array frame_history_gpu;
+
+	Label *frames_per_second;
```
```suggestion
	Label *frames_per_second = nullptr;
```
For all
